stdlib/seq.mli

```diff
@@ -13,97 +13,807 @@
+    drawn synchronously from the sequences [xs] and [ys] satisfies [p x y].
+
+    If the sequences [xs] and [ys] have different lengths, then
+    iteration must stop as soon as one sequence is exhausted;
+    the excess elements in the other sequence are ignored.
+
+    At least one of the sequences [xs] and [ys] must be finite.
+
+    [exists2 p xs ys] is equivalent to [exists (fun b -> b) (map2 p xs ys)].
+
+    @since 4.14 *)
+
+val equal: ('a -> 'b -> bool) -> 'a t -> 'b t -> bool
+(** Provided the function [eq] defines an equality on elements,
+    [equal eq xs ys] determines whether the sequences [xs] and [ys]
+    are pointwise equal.
+
+    At least one of the sequences [xs] and [ys] must be finite.
+
+    @since 4.14 *)
+
+val compare : ('a -> 'b -> int) -> 'a t -> 'b t -> int
+(** Provided the function [cmp] defines a preorder on elements,
+    [compare cmp xs ys] compares the sequences [xs] and [ys]
+    according to the lexicographic preorder.
+
+    For more details on comparison functions, see {!Array.sort}.
+
+    At least one of the sequences [xs] and [ys] must be finite.
+
+    @since 4.14 *)
+
+(** {1 Constructing sequences} *)
+
+(** The functions in this section are lazy: that is, they return sequences
+    whose elements are computed only when demanded. *)
 
 val empty : 'a t
-(** The empty sequence, containing no elements. *)
+(** [empty] is the empty sequence.
+    It has no elements. Its length is 0. *)
 
 val return : 'a -> 'a t
-(** The singleton sequence containing only the given element. *)
+(** [return x] is the sequence whose sole element is [x].
+    Its length is 1. *)
 
 val cons : 'a -> 'a t -> 'a t
-(** [cons x xs] is the sequence containing the element [x] followed by
-    the sequence [xs] @since 4.11 *)
+(** [cons x xs] is the sequence that begins with the element [x],
+    followed with the sequence [xs].
+
+    Writing [cons (f()) xs] causes the function call [f()]
+    to take place immediately. For this call to be delayed until the
+    sequence is queried, one must instead write
+    [(fun () -> Cons(f(), xs))].
 
-val append : 'a t -> 'a t -> 'a t
-(** [append xs ys] is the sequence [xs] followed by the sequence [ys]
     @since 4.11 *)
 
+val init : int -> (int -> 'a) -> 'a t
+(** [init n f] is the sequence [f 0; f 1; ...; f (n-1)].
+
+    [n] must be nonnegative.
+
+    [init n f] is equivalent to [map f (up 0 n)].
+
+    @raise Invalid_argument if [n] is negative.
+    @since 4.14 *)
+
+val unfold : ('b -> ('a * 'b) option) -> 'b -> 'a t
+(** [unfold] constructs a sequence
+    out of a step function and an initial state.
+
+    If [f u] is [None] then
+    [unfold f u] is the empty sequence.
+    If [f u] is [Some (x, u')] then
+    [unfold f u] is the nonempty sequence [cons x (unfold f u')].
+
+    For example, [unfold (function [] -> None | h :: t -> Some (h, t)) l]
+    is equivalent to [List.to_seq l].
+
+    @since 4.11 *)
+
+val repeat : 'a -> 'a t
```

Comment:
Nitpicking, but maybe `const` is a more natural name? (as in "constant sequence")

Reply:
Personally  I would be tempted to read `Seq.const v` as `Seq.return v`. 

Now when I read the signature I thought why isn't that `'a t -> 'a t` the answer is that this is provided by `cycle`. So `cycle` is for sequences and `repeat` for ground values. This seems in par with haskell terminology. 

(I would have prefered `always` or `forever` for this, but maybe it's not worth deviating from pre-existing names)

Reply:
Fun fact: `repeat` is the `pure/return` operation of the applicative whose product is the parallel/zipping product.